### ISSUE TITLE
AudioDriverWASAPI error while using editor, no audio feature used

### ISSUE THREAD

**Issue by Toshiwoz** [godotengine/godot#18732]
**Godot version:**
3.0.3 RC1

**OS/device including version:**
Windows 10, not related to video, apparently.

**Issue description:**
I have used the editor for a while and never got this error before, I haven't even used the audio features, yet this error message seems to be related to the audio:
```
ERROR: AudioDriverWASAPI::thread_func: WASAPI: GetCurrentPadding error
   At: drivers\wasapi\audio_driver_wasapi.cpp:533
```

**Steps to reproduce:**
Unkown

**Minimal reproduction project:**
unkonwn

**Comment by marcelofg55** [comment]
@Toshiwoz Hi, it happens everytime? Whats your soundcard?

**Comment by Toshiwoz** [comment]
Nope,
it started recently, strangely maybe, when I started using Netflix (it
sounds weird, but that aplication does strange things with the audio, it
does not allow to control the volume).

The sound card is an integrated soundcard, my pc is an old EEEBOX (with
intel atom), and I use HDMI + TV (with it's audio device)

2018-05-09 8:03 GMT-05:00 Marcelo Fernandez <notifications@github.com>:

**Comment by marcelofg55** [comment]
So if you don't use Netflix there is no problem? If so then it's probably requesting exclusive access to the soundcard.

**Comment by Toshiwoz** [comment]
Yeah, I can confirm that now, I opened Godot editor, my current project. No error.
Then I open Netflix, the moment I start playing a video, Godot starts to log that error.
I wonder why the Netflix app does that strange thing (should I report that to them then?).

**Comment by marcelofg55** [comment]
Yes, you could try to contact them since we can't do much if an external application requests WASAPI exclusive mode. They could switch to shared mode and that will solve the issue.

**Comment by Toshiwoz** [comment]
I've been told it's how it is suposed to work, by the netflix support, lol..
Should I close this?

**Comment by akien-mga** [comment]
I guess we could try to improve the error message, but we can't prevent thirdparty apps from abusing the audio driver :)

@marcelofg55 Would there be an easy way to catch this source of error and print an understandable error?

**Comment by marcelofg55** [comment]
Yeah, we should definitely improve the error messages for this so that it's more clear. But I think we won't be able to tell if another application that uses exclusive mode is actually the problem, afaik WASAPI funtions simply return AUDCLNT_E_DEVICE_INVALIDATED when this happens.

**Comment by OgierM** [comment]
I'm getting this error too; I am not using Netflix nor listening to any music.
`ERROR: thread_func: WASAPI: GetCurrentPadding error
        At: drivers/wasapi/audio_driver_wasapi.cpp:533`

**Comment by marcelofg55** [comment]
> I'm getting this error too; I am not using Netflix nor listening to any music.
ERROR: thread_func: WASAPI: GetCurrentPadding error
At: drivers/wasapi/audio_driver_wasapi.cpp:533

Happens every time you run godot? Or is it random? And what soundcard you're using?

**Comment by Toshiwoz** [comment]
It may happen also with any other program that takes full control of the sound card.
What I did was to remove the exclusive control of the program over the audio card... I don't remember where the setting is, there, in the image, try by unchecking "Allow applications to take exclusive control of this device".
![imagen](https://user-images.githubusercontent.com/21297356/44949884-2348b280-ae00-11e8-808b-ac7715a53913.png)

**Comment by OgierM** [comment]
Removing the exclusive mode rights stopped this behavior, I think, it's def not happening anymore rn and that's the only thing I've changed

**Comment by matthieudesprez** [comment]
Thanks, it did fix the error for me as well.

To access the settings on windows 10:
* right click the sound icon on the bottom right
* click on Sounds
* go on Playback tab
* select your audio device and click on Properties
* then you can go on Advanced tab and disable exclusive mode

**Comment by jitspoe** [comment]
Also getting this.  Windows 8.1.  Certainly happens opening some sort of sound application like Sound Forge, but I've had it happen pretty regularly just doing basic stuff like web browsing, though tough to say exactly what caused it.  I often have Godot open in the background and come back to see it spamming this nonstop.  Updated line number on 3.1.1:
```
 ERROR: thread_func: WASAPI: GetCurrentPadding error
   At: drivers/wasapi/audio_driver_wasapi.cpp:646
```
Might be nice to, at the very least, avoid spamming the error.  Best case would be to attempt to grab access of to the sound device again, maybe after a short delay.

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in [Godot 3.2.3 rc4](https://godotengine.org/article/release-candidate-godot-3-2-3-rc-4) or any later release?

**Comment by jitspoe** [comment]
Can confirm it still happens in 3.2.3 rc4.  I can repro it by simply opening godot to the project manager then launching VEGAS Pro, Sound Forge, or any other program that tries to take exclusive control of the sound card.

**Comment by kamtar** [comment]
somebody has a hot-fix for this? I would like to listen to Tidal the way I'm used to without rendering Godot unusable.

**Comment by elsandkls** [comment]
I'm running Godot Engine **v4.2.1.stable**. I'm using **VoiceMeter (Banana)** on my dev box to handle sound routing and device inputs/outputs. I just rebuilt the system and two days ago before I installed VoiceMeter I did not get this error in Godot, Yesterday I installed VoiceMeter and finished reconfiguring my system, and noticed this error has started. Is there a way to tell Godot Engine to cede control of audio to VoiceMeter? It just needs to output to the standard default output, without controlling it. VoiceMeter will take care of routing it to the device I want to hear it on.

Editor Errors:
```
E 0:00:01:0281   audio_device_init: WASAPI: Initialize failed with error 0xffffffff8889000a.
  <C++ Error>    Condition "hr != ((HRESULT)0x00000000)" is true. Returning: ERR_CANT_OPEN
  <C++ Source>   drivers/wasapi/audio_driver_wasapi.cpp:387 @ audio_device_init()
  
  E 0:00:01:0281   init: WASAPI: init_output_device error
  <C++ Source>   drivers/wasapi/audio_driver_wasapi.cpp:562 @ init()

E 0:00:01:0281   start: WASAPI: Start failed
  <C++ Source>   drivers/wasapi/audio_driver_wasapi.cpp:937 @ start()

E 0:00:01:0282   thread_func: WASAPI: GetCurrentPadding error
  <C++ Source>   drivers/wasapi/audio_driver_wasapi.cpp:808 @ thread_func()
--- and this last error just occurs continuously until I close the program.
```
I've read several cases, including older ones for Godot 3. It looks like Godot is supposed to share the audio control? 
Are there settings that I have missed that ensure this is behavior? 

In case you'd like to replicate the error the VoiceMeter software is available for free with donations recommended if you find it useful.  https://vb-audio.com/Voicemeeter/

**Comment by dmhagar** [comment]
There is no reason for Godot to literally log HUNDREDS of  thousands of these errors when this is happening.

A single, concise error message would be acceptable.

**Comment by allenwp** [comment]
I've created PR  #100116 to address the error spam in the Godot editor. If anyone is available for testing this fix, you can download an editor build from the [artifacts on the checks page of the PR](https://github.com/godotengine/godot/pull/100116/checks).

**Comment by elvisish** [comment]
> I've created PR [#100116](https://github.com/godotengine/godot/pull/100116) to address the error spam in the Godot editor. If anyone is available for testing this fix, you can download an editor build from the [artifacts on the checks page of the PR](https://github.com/godotengine/godot/pull/100116/checks).

This build seems to fix the spam problem when I use Realtek ASIO in Reaper at the same time as having Godot running. I get these errors when I start Reaper:
```swift
  drivers\wasapi\audio_driver_wasapi.cpp:780 - WASAPI: GetBufferSize error
  drivers\wasapi\audio_driver_wasapi.cpp:851 - WASAPI: Current output_device invalidated, closing output_device
  drivers\wasapi\audio_driver_wasapi.cpp:987 - WASAPI: Start failed
  drivers\wasapi\audio_driver_wasapi.cpp:780 - WASAPI: GetBufferSize error
  drivers\wasapi\audio_driver_wasapi.cpp:851 - WASAPI: Current output_device invalidated, closing output_device
  drivers\wasapi\audio_driver_wasapi.cpp:987 - WASAPI: Start failed
  drivers\wasapi\audio_driver_wasapi.cpp:780 - WASAPI: GetBufferSize error
  drivers\wasapi\audio_driver_wasapi.cpp:851 - WASAPI: Current output_device invalidated, closing output_device
  drivers\wasapi\audio_driver_wasapi.cpp:987 - WASAPI: Start failed
  drivers\wasapi\audio_driver_wasapi.cpp:780 - WASAPI: GetBufferSize error
  drivers\wasapi\audio_driver_wasapi.cpp:851 - WASAPI: Current output_device invalidated, closing output_device
```
But that's it. If I try to load Reaper with Realtek ASIO using regular 4.4.dev7, it spams until it hangs and I have to terminate it through Task Manager. I never had this problem with 3.5.

**Comment by goblinJoel** [comment]
I'd like to request @allenwp 's error spam PR  be done or cherry-picked for the next 3.x bugfix as well. (Not sure who to tag.) It's a more serious problem than it sounds.

**Gigs of logs, oh my!** I made a very simple time-tracking program for Windows in Godot 3.6.1, but today it just refused to launch for the user I made it for. Running the program apparently did nothing. We checked task manager and found a Godot process using over 300MB/s of disk! We killed it and tried it again with the same result. When we checked the app's logs, there were multiple multi-gigabyte logs it had created, spamming this error.

I re-exported it with debug off (but log files still enabled) to see if that would matter. It launched fine, but partway through the day it started doing this again. The program acts fine, but if we check the log files, we see the latest is 2 GB and climbing.

_When logging is enabled on an exported project, this error very rapidly creates multi-gigabyte log files and pounds the system storage drive._ (Even if the program doesn't use sound.)

For now, that project is tiny enough I'm upgrading it to 4.x, but I'm worried my much larger game project (which is stuck on 3.x) will either have to go without logs or risk filling my players' system drives without warning. We aren't sure what program even grabbed exclusive audio.

**Comment by allenwp** [comment]
Thanks for the ping! I'll coordinate with @lawnjelly about a cherry pick or a more custom fix for 3.x

**Comment by lawnjelly** [comment]
Yeah this didn't appear on the production radar:
* It's super old
* It's not marked in the 3.x / 3.7 milestone
* The PR was never marked as cherry pick 3.x even though it mentions a 3.x issue

So feel free to ping me or just mark the PR as cherry pick 3.x, or make sure any relevant issues are marked with a milestone. 👍
